packages/web-components/src/text-input/text-input.spec.ts

```diff
@@ -0,0 +1,636 @@
+                </fluent-text-input>
+            `;
+    });
+
+    await expect(label).toHaveClass(/label__hidden/);
+  });
+
+  test('should hide the label when end content is provided', async () => {
+    await root.evaluate(node => {
+      node.innerHTML = /* html */ `
+                <fluent-text-input>
+                    <div slot="end"></div>
+                </fluent-text-input>
+            `;
+    });
+
+    await expect(label).toHaveClass(/label__hidden/);
+  });
+
+  test('should hide the label when start and end content are provided', async () => {
+    await root.evaluate(node => {
+      node.innerHTML = /* html */ `
+                <fluent-text-input>
+                    <div slot="start"></div>
+                    <div slot="end"></div>
+                </fluent-text-input>
+            `;
+    });
+
+    await expect(label).toHaveClass(/label__hidden/);
+  });
+
+  test('should hide the label when space-only text nodes are slotted', async () => {
+    await root.evaluate(node => {
+      node.innerHTML = `<fluent-text-input>\n \n</fluent-text-input>`;
+    });
+
+    await expect(element).toHaveText(/\n\s\n/);
+
+    await expect(label).toHaveClass(/label__hidden/);
+  });
+
+  test('should fire a `change` event when the internal control emits a `change` event', async () => {
+    await root.evaluate(node => {
+      node.innerHTML = /* html */ `
+                <fluent-text-input></fluent-text-input>
+            `;
+    });
+
+    const [wasChanged] = await Promise.all([
+      element.evaluate(
+        node =>
+          new Promise(resolve => {
+            node.addEventListener('change', () => resolve(true));
+          }),
+      ),
+      control.evaluate(node => {
+        node.dispatchEvent(
+          new Event('change', {
+            key: 'a',
+          } as EventInit),
+        );
+      }),
+    ]);
+
+    expect(wasChanged).toBeTruthy();
+  });
+
+  test.describe('with a type of `password`', () => {
+    test('should report invalid validity when the `value` property is an empty string and `required` is true', async () => {
+      await root.evaluate(node => {
+        node.innerHTML = /* html */ `
+                    <fluent-text-input type="password" required></fluent-text-input>
+                `;
+      });
+
+      expect(await element.evaluate<boolean, TextInput>(node => node.validity.valueMissing)).toBeTruthy();
+    });
+
+    test('should report valid validity when the `value` property is a string that is non-empty and `required` is true', async () => {
+      await root.evaluate(node => {
+        node.innerHTML = /* html */ `
+                    <fluent-text-input type="password" required value="some-value"></fluent-text-input>
+                `;
+      });
+
+      expect(await element.evaluate<boolean, TextInput>(node => node.validity.valueMissing)).toBeFalsy();
+    });
+
+    test('should report valid validity when `value` is empty and `minlength` is set', async () => {
```

Comment:
What is this even?
Value: empty
Min length: 1
This means that the value is LESS than the minimum, and that's valid??

Reply:
The test is essentially asserting that the text input should report valid validity even when it's empty, despite having a minlength requirement. This might seem counterintuitive, but the HTML5 validation specifications allow inputs with minlength constraints to be valid when they are empty. The minlength constraint is only enforced when there is input in the field, not when it is empty.

So, the test is ensuring that the fluent-text-input component follows this specification correctly.